L1Regularization will throw NullPointerException

Caused by: java.lang.NullPointerException
	at com.intel.analytics.bigdl.optim.L1L2Regularizer.accL1Regularization(Regularizer.scala:138)
	at com.intel.analytics.bigdl.optim.L1L2Regularizer.accL1L2Regularization(Regularizer.scala:118)
	at com.intel.analytics.bigdl.optim.L1L2Regularizer.accRegularization(Regularizer.scala:98)
	at com.intel.analytics.bigdl.nn.SparseLinear.accGradParameters(SparseLinear.scala:123)
	at com.intel.analytics.bigdl.nn.SparseLinear.accGradParameters(SparseLinear.scala:44)
	at com.intel.analytics.bigdl.nn.abstractnn.AbstractModule.backward(AbstractModule.scala:306)
	at com.intel.analytics.bigdl.nn.Sequential.backward(Sequential.scala:83)
	at com.intel.analytics.bigdl.optim.DistriOptimizer$$anonfun$8$$anonfun$9$$anonfun$apply$2.apply$mcI$sp(DistriOptimizer.scala:226)
	at com.intel.analytics.bigdl.optim.DistriOptimizer$$anonfun$8$$anonfun$9$$anonfun$apply$2.apply(DistriOptimizer.scala:216)
	at com.intel.analytics.bigdl.optim.DistriOptimizer$$anonfun$8$$anonfun$9$$anonfun$apply$2.apply(DistriOptimizer.scala:216)
	at com.intel.analytics.bigdl.utils.ThreadPool$$anonfun$1$$anon$4.call(ThreadPool.scala:112)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)

--- a/spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Regularizer.scala
+++ b/spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Regularizer.scala
@@ -134,11 +134,14 @@ class L1L2Regularizer[T: ClassTag](
     gradParameter: Tensor[T],
     scale: Double
   ): Unit = {
-    if (alpha != 0 && scale != 0) gradParameter.add(ev.fromType(alpha*scale),
-      l1SignBuffer.resizeAs(parameter).copy(parameter).sign())
+    if (alpha != 0 && scale != 0) {
+      if (null == l1SignBuffer) l1SignBuffer = Tensor()
+      gradParameter.add(ev.fromType(alpha*scale),
+        l1SignBuffer.resizeAs(parameter).copy(parameter).sign())
+    }
   }
 
-  @transient private val l1SignBuffer = Tensor()
+  @transient private var l1SignBuffer: Tensor[T] = null
 
   /**
    * Accumulates the gradient of the l2 regularization of `parameter`

--- a/spark/dl/src/test/scala/com/intel/analytics/bigdl/optim/DistriOptimizerSpec.scala
+++ b/spark/dl/src/test/scala/com/intel/analytics/bigdl/optim/DistriOptimizerSpec.scala
@@ -568,4 +568,32 @@ class DistriOptimizerSpec extends FlatSpec with Matchers with BeforeAndAfter {
     val result2 = model.forward(input2).asInstanceOf[Tensor[Double]]
     result2(Array(1)) should be(1.0 +- 5e-2)
   }
+
+  "Train with L1Regularization" should "work properly in DistriOptimizer" in {
+    LoggerFilter.redirectSparkInfoLogs()
+    Logger.getLogger("com.intel.analytics.bigdl.optim").setLevel(Level.INFO)
+    Logger.getLogger("com.intel.analytics.bigdl").setLevel(Level.INFO)
+
+    RandomGenerator.RNG.setSeed(10)
+    val logdir = com.google.common.io.Files.createTempDir()
+    val mm = Sequential[Double]().add(Linear(4, 2,
+      wRegularizer = L1Regularizer(1), bRegularizer = L1Regularizer(1)))
+      .add(Sigmoid())
+      .add(Linear(2, 1))
+      .add(Sigmoid())
+    mm.getParameters()._1.fill(0.125)
+    val optimizer = new DistriOptimizer[Double](
+      _model = mm,
+      dataset = dataSet,
+      criterion = new MSECriterion[Double]()
+    )
+
+    val optimMethod = new SGD[Double](learningRate = 20.0)
+
+    optimizer.setOptimMethod(optimMethod)
+      .setEndWhen(Trigger.severalIteration(10))
+    optimizer.setValidation(Trigger.everyEpoch, dataSet,
+      Array(new Top1Accuracy[Double]()))
+    val model = optimizer.optimize()
+  }
 }

